### DISCUSSION TITLE
Setting the PK of a foreign key: corresponding object is not linked. Where is that discussed in the docs?

### DISCUSSION THREAD

**Discussion by taigit** [sqlalchemy/sqlalchemy#11477] [Usage Questions] [answered]
During creating of an object, I set the PK of a foreign key and expect the corresponding object link to be established.

It does not work unless I flush the session.

.

I'm trying to find the place in the docs where it says that.

I looked at

https://docs.sqlalchemy.org/en/20/tutorial/orm_related_objects.html#persisting-and-loading-relationships

https://docs.sqlalchemy.org/en/20/tutorial/orm_data_manipulation.html#inserting-rows-using-the-orm-unit-of-work-pattern

but could not find that statement.

.

I asked chat gpt 4o, but it says:

session.flush() is called again after setting the parent_id to ensure that the ID is updated in the database.

ie: it says that flush is updating the db. But what i am looking for is user level guidance about how to make my session info consistent and usable.

Is this in the docs? where?

**Comment by taigit** [comment]
What's the recommended solution for this situation?

Should I flush the session? Every time I want a consistent version of the session?

Should I avoid using the object ID and instead I grab the object and associate the object? This would avoid a write in the database.

Are there other options?

**Reply by zzzeek** [reply]
>  I set the PK of a foreign key and expect the corresponding object link to be established.

we have an extensive FAQ entry on this here: https://docs.sqlalchemy.org/en/20/faq/sessions.html#i-set-the-foo-id-attribute-on-my-instance-to-7-but-the-foo-attribute-is-still-none-shouldn-t-it-have-loaded-foo-with-id-7

**Comment by CaselIT** [comment]
Hi,

That's the expected behaviour, since add does not do io.

**Reply by zzzeek** [reply]
well to be clear if they are already assigning the PK in Python, the value is there, it could be transferred to the FK attributes on the other object without any SQL or IO.

**Reply by CaselIT** [reply]
I think the ask was to load the scalalr value of the relationship given the fk. Like in the example below

**Comment by taigit** [comment]
ok, that makes sense.  what's the recommendation for users?
.
.
here's a clarification of the situation that we are talking about:


### Example Code

```python
from sqlalchemy import create_engine, Column, Integer, String, ForeignKey
from sqlalchemy.orm import relationship, sessionmaker, declarative_base

Base = declarative_base()

class Parent(Base):
    __tablename__ = 'parent'
    id = Column(Integer, primary_key=True)
    name = Column(String)

class Child(Base):
    __tablename__ = 'child'
    id = Column(Integer, primary_key=True)
    parent_id = Column(Integer, ForeignKey('parent.id'))
    name = Column(String)

    parent = relationship('Parent', back_populates='children')

Parent.children = relationship('Child', order_by=Child.id, back_populates='parent')

# Setting up the engine and session
engine = create_engine('sqlite:///:memory:')
Base.metadata.create_all(engine)
Session = sessionmaker(bind=engine)
session = Session()

# Creating and committing a parent object
parent = Parent(name='Parent 1')
session.add(parent)
session.commit()  # Ensure the parent ID is generated

# Creating a child and associating with the parent
child = Child(name='Child 1', parent_id=parent.id)
session.add(child)

# Attempt to access the parent object from the child
# This access fails unless we flush the session first
# session.flush()  # Uncommenting this line makes the parent accessible

try:
    # Accessing the parent object
    print(child.parent)
except Exception as e:
    print(f'Error: {e}')

# Committing the session
session.commit()

# Verifying the relationship after commit
retrieved_child = session.query(Child).filter_by(id=child.id).first()
print(f'Retrieved Child: {retrieved_child}')
print(f'Child Name: {retrieved_child.name}, Parent ID: {retrieved_child.parent_id}, Parent Name: {retrieved_child.parent.name}')
```

### Question for Experts

I'm encountering an issue in SQLAlchemy when trying to create a child object and associate it with an existing parent object. Here is what I'm doing:

1. I have a `Parent` table and a `Child` table, with a foreign key relationship between them.
2. I create and commit a `Parent` object first.
3. Then I create a `Child` object and set its `parent_id` to the ID of the existing `Parent`.
4. After adding the `Child` object to the session, I try to access the `parent` attribute of the `Child`.

However, the `parent` attribute is not accessible unless I call `session.flush()` before accessing it. Here is the relevant code snippet:

```python
# Creating and committing a parent object
parent = Parent(name='Parent 1')
session.add(parent)
session.commit()  # Ensure the parent ID is generated

# Creating a child and associating with the parent
child = Child(name='Child 1', parent_id=parent.id)
session.add(child)

# Attempt to access the parent object from the child
# This access fails unless we flush the session first
# session.flush()  # Uncommenting this line makes the parent accessible

try:
    # Accessing the parent object
    print(child.parent)
except Exception as e:
    print(f'Error: {e}')

# Committing the session
session.commit()

# Verifying the relationship after commit
retrieved_child = session.query(Child).filter_by(id=child.id).first()
print(f'Retrieved Child: {retrieved_child}')
print(f'Child Name: {retrieved_child.name}, Parent ID: {retrieved_child.parent_id}, Parent Name: {retrieved_child.parent.name}')
```

### My Question

Why is it necessary to call `session.flush()` before accessing the `parent` attribute of the `Child` object? Is there a way to make the parent accessible without explicitly calling `session.flush()`? What is the best practice in this scenario to ensure the relationship is correctly established and accessible?

**Reply by zzzeek** [reply, marked as answer]
see the FAQ entry at https://docs.sqlalchemy.org/en/20/faq/sessions.html#i-set-the-foo-id-attribute-on-my-instance-to-7-but-the-foo-attribute-is-still-none-shouldn-t-it-have-loaded-foo-with-id-7

**Comment by taigit** [comment]
Ok, that's for the clarification.

From the user's pov, the usage summary seems to be: insert/update .vs. read:

if you are doing insert or update, you are on your own and don't rely on sqlalchemy relationship linking (for values that have been changed). If you want to use relationship linking, then you need to set it up correctly yourself by assigning the objects as relationship objects (eg: the parent object for this child object needs to be explicitly assigned to be object P).

if you are doing read only, then you can rely on sqlalchemy relationship linking.
